docs/arch/adr-2505-eventual-consistency.md

```diff
@@ -0,0 +1,20 @@
+# ADR 2505: Enable eventual consistency for maintaining search param indexes in sync with the Resource table
+Labels: [SQL](https://github.com/microsoft/fhir-server/labels/Area-SQL)
+
+## Context
+Efficiency of writes to FHIR SQL storage is higly dependent on duration of SQL transactions. Long SQL transactions cause significant blocking between parallel ingestion processes and reduce overall write throughput. This is usually observered for large hyperscale databases (>1TB). Currently FHIR SQL supports strong consistency when updates in the Resource table and the search param tables are included in the same SQL transaction. This makes SQL transactions long.
+
+## Decision
+We will implement eventual consistency for maintaining search param indexes. Updates of resources will still run in single SQL transaction as today (strong consistency), because they involve deletion of previous search param rows. For resource creates only, we will run all writes in separate SQL transactions, starting with insert into the Resource table. In case of success this is equivalent to current strong consistency behavior. In case of interruptions search param indexes might not be completely in sync with corresponding row in the Resource table. This will be addressed by the Transaction watchdog, which will roll transaction forward based on data committed in the Resource table.
```

Comment:
So the changes are?
1. Turning off transactions held from c# code
2. Import uses transaction table id on imported data
3. Watchdog ensures consistency and rolls transactions forward

Reply:
$import does not use C# transactions, so no code changes for $import. 

In future, if we want to extend eventual consistency usage for bundles and other API calls, we will not do this for cases where C# transactions are needed. These will work as they do today. For cases where all PUTs in bundle are aggregated into a single SQL call, we can use the same method $import will be using (new input parameter).

Today all writes, $import and any calls via API, are already using Transactions table, so there is no change here.

Today Transaction watchdog already deals with timeout transactions and rolls them forward, so no changes here.